atomic_reactor/plugins/exit_koji_import.py

```diff
@@ -255,10 +260,24 @@ def set_operators_metadata(self, extra, worker_metadatas):
                             'archive': OPERATOR_MANIFESTS_ARCHIVE,
                         },
                     }
-                    extra.update({'typeinfo': operators_typeinfo})
+                    extra.setdefault('typeinfo', {}).update(operators_typeinfo)
 
                     return  # only one worker can process operator manifests
 
+    def set_remote_sources_metadata(self, extra):
+        remote_source_result = self.workflow.prebuild_results.get(PLUGIN_RESOLVE_REMOTE_SOURCE)
+        if remote_source_result:
+            url = remote_source_result['annotations']['remote_source_url']
+            remote_source_typeinfo = {
+                KOJI_BTYPE_CONTAINER_FIRST: {
+                    'remote_source_url': url,
+                },
+            }
+            extra.setdefault('typeinfo', {}).update(remote_source_typeinfo)
+
+            # TODO: is setting it in the image metadata also needed?
```

Comment:
We do need the container-first metadata in the typeinfo to set the btype here. Should we use it to hold the remote_source_url metadata as above? In this case, should we keep the duplication below?

Reply:
I'm not sure that's right.

'typeinfo' is about the build, not about a specific output of the build:

> [build.extra.]typeinfo: A map whose **single entry** is the name of the content generator type, which is a map containing type-specific information for this build.

(from [docs](https://docs.pagure.org/koji/content_generator_metadata/), emphasis mine)

whereas outputs have:

> type: The type of the file. Log files should use “log”.

and no other types are specified -- seems like it's a freeform string. There's no typeinfo in the output's 'extra' map either.

Reply:
I am not sure how accurate is the documentation though...

As per [1], it seems several entries are processed in the typeinfo map. With [2] it seems a build type is set based on the information in the typeinfo field.

By looking at [3], it also seems that we need the btype in the typeinfo so koji will build (and display) URLs for the archives.

The reason we do not have typeinfo in most of OSBS build metadata is because most builds use only the `image` legacy type. Operator builds do have the typeinfo field as well since they have their own koji btype.

[1] https://pagure.io/koji/blob/master/f/hub/kojihub.py#_6092
[2] https://pagure.io/koji/blob/master/f/hub/kojihub.py#_4300
[3] https://pagure.io/koji/blob/master/f/www/kojiweb/index.py#_1124

Reply:
OK, thanks for finding the facts :-)

Reply:
I opened https://pagure.io/koji/pull-request/1917 to update those docs.